crates/svm-runtime/src/import.rs

```diff
@@ -1,51 +1,169 @@
+    pub unsafe fn inner(&self) -> &Context {
+        &*{ self.inner_env as *const Context }
+    }
+
+    pub unsafe fn inner_mut(&self) -> &mut Context {
+        &mut *(self.inner_env as *const Context as *mut Context)
+    }
+
+    pub unsafe fn host_env<T>(&self) -> &T {
+        &*(self.host_env as *const T)
+    }
+}
+
+impl Drop for svm_env_t {
+    fn drop(&mut self) {
+        dbg!("dropping `svm_env_t`");
+    }
+}
 
-#[derive(Debug, PartialEq, Clone)]
-pub struct Import {
+impl From<*mut c_void> for &svm_env_t {
+    fn from(env: *mut c_void) -> Self {
+        unsafe { &*(env as *mut svm_env_t) }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ExternImport {
     pub name: String,
 
+    pub namespace: String,
+
     pub params: Vec<WasmType>,
 
     pub returns: Vec<WasmType>,
 
     pub func_ptr: *const c_void,
+
+    pub host_env: *const c_void,
 }
 
-impl Import {
-    pub fn to_wasmer(&self, ctx: Context) -> Export {
-        let params = to_wasmer_types(&self.params);
-        let returns = to_wasmer_types(&self.returns);
-        let signature = FunctionType::new(params, returns);
+impl ExternImport {
+    pub fn wasmer_export(&self, store: &Store, ctx: &mut Context) -> (Export, *const svm_env_t) {
+        unsafe {
+            // This code is almost a clone of the code here:
+            // https://github.com/wasmerio/wasmer/blob/7847acaae1e7a0eade13b65def1f3feeac95efd7/lib/c-api/src/wasm_c_api/externals/func.rs#L86
 
-        // TODO: needs to free `ctx`.
-        let boxed_ctx = Box::new(ctx);
-        let vmctx = Box::into_raw(boxed_ctx) as *mut _ as *mut VMContext;
+            let func_ty = self.wasmer_function_ty();
 
-        let func = ExportFunction {
-            address: self.func_ptr as _,
-            vmctx,
-            signature,
-            kind: VMFunctionKind::Static,
-            call_trampoline: None,
-        };
+            let callback: wasm_func_callback_with_env_t = std::mem::transmute(self.func_ptr);
+
+            let num_rets = func_ty.results().len();
+
+            let inner_callback =
+                move |env: &mut *mut c_void, args: &[Val]| -> Result<Vec<Val>, RuntimeError> {
+                    let processed_args: wasm_val_vec_t = args
+                        .into_iter()
+                        .map(TryInto::try_into)
+                        .collect::<Result<Vec<wasm_val_t>, _>>()
+                        .expect("Argument conversion failed")
+                        .into();
+
+                    let zero = wasm_val_t::try_from(Val::I64(0)).unwrap();
+                    let mut results: wasm_val_vec_t = vec![zero; num_rets].into();
+
+                    let trap = callback(*env, &processed_args, &mut results);
+
+                    if !trap.is_null() {
+                        let trap: Box<wasm_trap_t> = Box::from_raw(trap);
+
+                        /// TODO: we want access to `trap.inner`
+                        /// (this field has visibility of `pub(crate)`)
```

Comment:
So we don't have a solution for this yet?

Reply:
no, we need to ask the wasmer team to change that